src/phase0/state_transition/epoch_processing.rs

```diff
@@ -86,6 +88,69 @@ pub fn get_matching_target_attestations<
         .filter(move |a| a.data.target.root == *block_root))
 }
 
+pub fn _get_matching_head_attestations<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    _state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    _context: &Context,
+) -> Result<(), Error> {
+    Ok(())
+}
+
+pub fn get_unslashed_attesting_indices<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    attestations: &[PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>],
+    context: &Context,
+) -> Result<HashSet<ValidatorIndex>, Error> {
+    let output = HashSet::new();
+    for a in attestations.iter() {
+        output.union(&get_attesting_indices(
+            state,
+            &a.data,
+            &a.aggregation_bits,
+            context,
+        )?);
+    }
+    Ok(output
+        .into_iter()
+        .filter(|&i| !state.validators[i].slashed)
+        .collect::<HashSet<ValidatorIndex>>())
```

Comment:
this is another place we could consider removing the intermediate allocations by dropping `collect` and considering how we build the set `output` but i think for now this is fine -- there are bigger hot spots we will hit and i don't know if we will hit this level of optimization in this project tbh

Reply:
Cool...I'll leave this one alone for now.

Reply:
If needed -- this is a very minor change with the latest commit. Like <5 mins of work, so I'll default to you if we want an "existential" type here to keep things consistent